src/blockrelay/compactblock.cpp

```diff
@@ -0,0 +1,1216 @@
+#include "expedited.h"
+#include "hash.h"
+#include "main.h"
+#include "net.h"
+#include "parallel.h"
+#include "policy/policy.h"
+#include "pow.h"
+#include "random.h"
+#include "requestManager.h"
+#include "streams.h"
+#include "timedata.h"
+#include "txadmission.h"
+#include "txmempool.h"
+#include "txorphanpool.h"
+#include "util.h"
+#include "utiltime.h"
+#include "validation/validation.h"
+
+
+static bool ReconstructBlock(CNode *pfrom, const bool fXVal, int &missingCount, int &unnecessaryCount);
+
+
+uint64_t GetShortID(const uint64_t &shorttxidk0, const uint64_t &shorttxidk1, const uint256 &txhash)
+{
+    static_assert(CompactBlock::SHORTTXIDS_LENGTH == 6, "shorttxids calculation assumes 6-byte shorttxids");
+    return SipHashUint256(shorttxidk0, shorttxidk1, txhash) & 0xffffffffffffL;
+}
+
+#define MIN_TRANSACTION_SIZE (::GetSerializeSize(CTransaction(), SER_NETWORK, PROTOCOL_VERSION))
+
+CompactBlock::CompactBlock(const CBlock &block, const CRollingFastFilter<4 * 1024 * 1024> *inventoryKnown)
+    : nonce(GetRand(std::numeric_limits<uint64_t>::max())), header(block)
+{
+    FillShortTxIDSelector();
+
+    if (block.vtx.empty())
+        throw std::invalid_argument(__func__ + std::string(" expects coinbase tx"));
+
+    //< Index of a prefilled tx is its diff from last index.
+    size_t prevIndex = 0;
+    prefilledtxn.push_back(PrefilledTransaction{0, *block.vtx[0]});
+    for (size_t i = 1; i < block.vtx.size(); i++)
+    {
+        const CTransaction &tx = *block.vtx[i];
+        if (inventoryKnown && !inventoryKnown->contains(tx.GetHash()))
+        {
+            prefilledtxn.push_back(PrefilledTransaction{static_cast<uint16_t>(i - (prevIndex + 1)), tx});
+            prevIndex = i;
+        }
+        else
+        {
+            shorttxids.push_back(GetShortID(tx.GetHash()));
+        }
+    }
+}
+
+void CompactBlock::FillShortTxIDSelector() const
+{
+    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
+    stream << header << nonce;
+    CSHA256 hasher;
+    hasher.Write((unsigned char *)&(*stream.begin()), stream.end() - stream.begin());
+    uint256 shorttxidhash;
+    hasher.Finalize(shorttxidhash.begin());
+    shorttxidk0 = shorttxidhash.GetUint64(0);
+    shorttxidk1 = shorttxidhash.GetUint64(1);
+}
+
+uint64_t CompactBlock::GetShortID(const uint256 &txhash) const
+{
+    return ::GetShortID(shorttxidk0, shorttxidk1, txhash);
+}
+
+void validateCompactBlock(const CompactBlock &cmpctblock)
+{
+    if (cmpctblock.header.IsNull() || (cmpctblock.shorttxids.empty() && cmpctblock.prefilledtxn.empty()))
+        throw std::invalid_argument("empty data in compact block");
+
+    int32_t lastprefilledindex = -1;
+    for (size_t i = 0; i < cmpctblock.prefilledtxn.size(); i++)
+    {
+        if (cmpctblock.prefilledtxn[i].tx.IsNull())
+            throw std::invalid_argument("null tx in compact block");
+
+        lastprefilledindex += cmpctblock.prefilledtxn[i].index + 1; // index is a uint16_t, so cant overflow here
```

Comment:
does this mean compact blocks are only good for 65k tx?

Reply:
Now that I think about @dagur modified xt and ABC code so that it uses an uint32 (or 64) for index

Reply:
sorry @Dagur, I menat to tag @dagurval.